[fix] Sync subgraph node title changes with breadcrumb navigation

christian-byrne opened the pull request:
Fixes an issue where the breadcrumb navigation didn't update when subgraph node titles were changed via double-click editing. The fix ensures title changes are synchronized across all data structures for proper reactivity and persistence.


https://github.com/user-attachments/assets/cb357502-f263-49b1-898e-8e5acb85efb6

closes https://github.com/Comfy-Org/ComfyUI_frontend/issues/4178#

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4394-fix-Sync-subgraph-node-title-changes-with-breadcrumb-navigation-22b6d73d3650818eb4b7cf40de47c646) by [Unito](https://www.unito.io)

christian-byrne committed:
[fix] sync subgraph node title changes with breadcrumb navigation

- Add useSubgraphTitleSync composable to handle title synchronization
- Update TitleEditor to sync both runtime and exported subgraph names
- Ensure breadcrumb updates reactively when subgraph node titles change
- Add comprehensive tests for the new functionality

christian-byrne committed:
[test] add browser test for subgraph breadcrumb title synchronization

- Test verifies breadcrumb updates when subgraph node title is changed
- Uses existing nested-subgraph.json workflow for testing
- Follows established test patterns for node interactions

webfiltered reviewed (changes requested):
Current design is directly editing the internal state of `changeTracker`, bypassing the API.  Is there a limitation that requires this?  We should not need something this low-level for a simple title change.

christian-byrne committed:
[fix] simplify subgraph breadcrumb title sync implementation

- Remove unnecessary composable and inline simple sync logic directly in TitleEditor
- Only sync runtime subgraph.name for breadcrumb reactivity (not exported definitions)
- Addresses code review feedback about bypassing APIs and manual state synchronization
- Graph serialization handles persistence automatically without manual intervention

webfiltered reviewed (changes requested):
Looks like tests failed.

christian-byrne commented on src/locales/en/settings.json:
```diff
@@ -322,6 +322,7 @@
   },
   "Comfy_UseNewMenu": {
     "name": "Use new menu",
+    "tooltip": "Menu bar position. On mobile devices, the menu is always shown at the top.",
```
Haha... mistakes were made.

This tooltip was just changed in recent PR but the PR was made when the `update-locales` action was exiting early because pre-commit was still failing excessively.

christian-byrne commented on src/components/graph/TitleEditor.vue:
```diff
@@ -41,7 +41,15 @@ const previousCanvasDraggable = ref(true)
+
+    // If this is a subgraph node, sync the runtime subgraph name for breadcrumb reactivity
+    const target = titleEditorStore.titleEditorTarget
+    if (target instanceof LGraphNode && target.isSubgraphNode?.()) {
```
Apparently this rule will catch these https://typescript-eslint.io/rules/no-unnecessary-condition/

Should we add it?

christian-byrne replied:
I'm not sure, I just read it on stackoverflow.

christian-byrne replied:
They have it in their examples on the docs for the rule, but i can verify before PR

<img width="954" height="188" alt="Selection_1684" src="https://github.com/user-attachments/assets/7616a97b-3b40-4c3b-a947-93b442a9b51c" />

christian-byrne replied:
I tested adding the rule and there's 391 linting errors violating it in current code.

They are mostly auto-fixable, but then I realized: in many places, OUR code can definitively say that something is never null, but the code from extensions is not compiled against our types. So, the null checks are unnecesary from naive perspective of our compilation stage, but often added in reality because of custom nodes.

I wonder if there's realistically anything we can do about that? WDYT?

christian-byrne committed:
increase timeout to 10s on breacrumb render

christian-byrne committed:
use new menu in browser test beforeEach

christian-byrne commented:
Ah, tests were failing because of the classic "creating new browser tests file and forgetting to set `UseNewMenu: Top`."

I need to add a line in CLAUDE.md and cursorrules for that.

christian-byrne commented:
Going to add the update to nodeDefStore in a followup PR.
